### PR TITLE
Allow matching of "any" attributes in find_matching_product_variation…

### LINKED ISSUE #23042: find_matching_product_variation() behaviour with "Any" variation attributes
**Describe the bug**
`find_matching_product_variation()` expects to receive a variable product instance, and an array of attributes so that it can find which of the individual variations matches those `match_attributes`.

It currently returns no results (return value of `0`) if you do not pass it a `match_attribute` (even empty) for an attribute where the variation is set to match "any" of the attribute values. 

E.g. Consider a variable product that differentiates on size attribute, but also has "colour" as an attribute used for variations, e.g.

Screenshot 2019-03-15 15 34 06

`find_matching_product_variation()` should be able to find the correct variation given *just* the size. However if fails if it is passed only a size.

```
// This returns 0
$data_store   = WC_Data_Store::load( 'product' );
$variation_id = $data_store->find_matching_product_variation(
    $variable_product,
    [
        'attribute_pa_size' => 'large',
    ]
);
```

```
// This returns the correctly variation ID (74 in our example)
$data_store   = WC_Data_Store::load( 'product' );
$variation_id = $data_store->find_matching_product_variation(
    $variable_product,
    [
        'attribute_pa_size' => 'large',
        'attribute_pa_colour' => '',
    ]
);
```

The problem appears to be this code here which returns 0 if an attribute is used for variations on a product and no attribute value is passed into the match_attributes array:
https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-data-store-cpt.php#L1068

I've come up with a potential fix [PR to follow] that seems to work for me, but I'm not 100% sure the logic is correct so it could definitely do with a review.

### PR DESCRIPTION
**Pull request by mikejolley**
Based on #23043 but corrects "any" logic.

Closes #23042

Force `woocommerce_ajax_variation_threshold` to 1 then test a variation with a required attribute, and an 'any' attribute and ensure it matches. 

@leewillis77 

Lee, any instructions to add to test the thing this is fixing? Possible to make a unit test?

### CODE REVIEW

**Comment by leewillis77** [comment]
Hi @mikejolley 

Thanks for the work on this. This PR looks good to me.

I've got a unit test for this (that fails with master, and passes with his PR). Patch attached since I don't think I can push to your PR branch - let me know if you want that raised as a separate PR for ease. 

[unit-test.txt](https://github.com/woocommerce/woocommerce/files/2978416/unit-test.txt)

**However**, that unit test also passes with my original PR so they obviously don't cover everything since you had a case that didn't work with my changes?

File: tests/unit-tests/product/data-store.php [hunk 1]
```diff
@@ -910,4 +913,52 @@ public function test_variable_create_all_product_variations_limits() {
...
+		$this->assertEquals( $children[0], $match );
+
+		$match = $data_store->find_matching_product_variation(
+			$product,
+			array(
+				'attribute_pa_size'   => 'small',
+				'attribute_pa_colour' => 'red',
+			)
+		);
+		$this->assertEquals( $children[0], $match );
```
Comment (peterfabian) [inline tests/unit-tests/product/data-store.php]: Great that this change comes in with unit tests! 
Should this test also explore other eventualities, like if only the colour is defined or if non-existing values are passed to the function?

File: assets/css/admin.scss [hunk 2]
```diff
@@ -3929,8 +3929,6 @@ img.help_tip {
 			top: 20px;
 		}
 
-		woocommerce-help-tip
-
```
Comment (peterfabian) [inline assets/css/admin.scss]: Is this something coming from the merge? Seems unrelated to the rest of the PR?

**Review by peterfabian — CHANGES_REQUESTED** [review]
I tested this quite a bit today, especially to check that:
- variation can be added to cart (to test function call from `add_to_cart_handle_variable`)
- variation is selected properly based on attributes on the front end (if the Ajax call to `get_variation` gets triggered, as that calls the updated function as well)

It seems to me that functionality-wise, all seems to behave correctly.

Similarly to unit tests, I created a product with size (S, M, L) and a couple of colours (green, blue, red). There were 3 variations, one for each size, colour is defined as Any.

Looking at how the function modified its behaviour, this is what I saw:

| param size  | param colour |  value on master |  valuer in this PR |
| :---: | :---: | :---: | :---: |
| "L"  | N/A  | 0 | correct var number |
| "L"  | ""  | correct var number | correct var number |
| "L"  | "color_that_doesnt_exist"  | var_number_X | var_number_X |
| N/A  | "green"  | 0 | 0 |
| ""  | "green"  | null | 0 |
| ""  | ""  | null | 0 |
| N/A  | N/A  | 0 | 0 |

Based on that, I think it _should be_ fine to give this one a go, although I feel a bit strange that asking for variation of size L and colour Turquoise if there is no Turquoise colour defined returns the variations assigned to size L could be a bit of a quirk, but it behaves consistently, so should be fine I think.
Other values that differ should all evaluate to _false_ and the first row is the desired difference.

I'm just unsure about the CSS change, that seems to me it does not belong here, so that's the only problem we should address, I think.

File: assets/css/admin.scss (hunk 2 above)
Comment (claudiosanches) [reply assets/css/admin.scss]: @peterfabian merge conflicts, still this line sounds like a bug anyway.

Comment (peterfabian) [reply assets/css/admin.scss]: @claudiosanches Can you please push an update? Don't want to add commits myself as a reviewer...

Comment (kloon) [reply assets/css/admin.scss]: In master that line is there but seems out of place https://github.com/woocommerce/woocommerce/blob/master/assets/css/admin.scss#L3932

Comment (kloon) [reply assets/css/admin.scss]: This is the commit that introduced the string https://github.com/woocommerce/woocommerce/commit/64c71cb7243e6bc52ad1c782ce01bd38fc55e673#diff-6b2f6833e79a750bf7f3a84223058d3eR3927 Which seems odd as that string should not be present. I think it would be fine to remove it from this PR, linting probably cleaned it up.

**Comment by kloon** [comment]
Added a couple more tests to test all the combinations in the table @peterfabian documented above.

File: assets/css/admin.scss (hunk 2 above)
Comment (peterfabian) [reply assets/css/admin.scss]: Thanks for investigating, @kloon. Let's keep the change as it is, then.

**Review by peterfabian — APPROVED** [review]
Should be good to go now.
